apis/Google.Cloud.Logging.Console/Google.Cloud.Logging.Console/GoogleCloudConsoleFormatter.cs

```diff
@@ -23,120 +23,195 @@
+            {
+                writer.WriteStartObject();
+                writer.WriteString(s_messagePropertyName, message);
+                writer.WriteString(s_categoryPropertyName, logEntry.Category);
+                writer.WriteString(s_severityPropertyName, GetSeverity(logEntry.LogLevel));
+                if (logEntry.Exception is object)
+                {
+                    writer.WriteString(s_exceptionPropertyName, ToInvariantString(logEntry.Exception));
+                }
+
+                MaybeWriteFormatParameters(writer, logEntry.State);
+                MaybeWriteScopeInformation(writer, scopeProvider);
+                writer.WriteEndObject();
+                writer.Flush();
+            }
+            // TODO: It would be lovely to have a more efficient way of writing the JSON to the console
+            // than writing bytes and converting those bytes back into a string - but for the moment,
+            // this is what we've got.
+            textWriter.WriteLine(Encoding.UTF8.GetString(output.GetBuffer(), 0, (int) output.Position));
         }
+    }
 
-        /// <summary>
-        /// Constructs a new formatter which uses the specified monitor to retrieve
-        /// options and watch for options changes.
-        /// </summary>
-        /// <param name="optionsMonitor">The monitor to observe for changes in options. Must not be null.</param>
-        public GoogleCloudConsoleFormatter(IOptionsMonitor<GoogleCloudConsoleFormatterOptions> optionsMonitor)
-            : this(optionsMonitor.CurrentValue)
+    private void MaybeWriteFormatParameters<TState>(Utf8JsonWriter writer, TState state)
+    {
+        // If we have format params and its more than just the original message add them.
+        if (state is IEnumerable<KeyValuePair<string, object>> formatParams &&
+            ContainsFormatParameters(formatParams))
         {
-            _optionsReloadToken = optionsMonitor.OnChange(options => _options = options);
+            writer.WritePropertyName(s_formatParametersPropertyName);
+            WriteKeyValuePairs(writer, formatParams);
         }
 
-        /// <inheritdoc />
-        public override void Write<TState>(in LogEntry<TState> logEntry, IExternalScopeProvider scopeProvider, TextWriter textWriter)
+        // Checks that fields is:
+        // - Non-empty
+        // - Not just a single entry with a key of "{OriginalFormat}"
+        // so we can decide whether or not to populate a struct with it.
+        static bool ContainsFormatParameters(IEnumerable<KeyValuePair<string, object>> fields)
         {
-            string message = logEntry.Formatter(logEntry.State, logEntry.Exception);
-            if (logEntry.Exception == null && message == null)
-            {
-                return;
-            }
-            // TODO: Use PooledByteBufferWriter when it's public...
-            using var output = new MemoryStream();
+            using (var iterator = fields.GetEnumerator())
```

Comment:
Wondering if this could be done with someting like

`fields.Any(pair => pair.Key != "{OriginalFormat}")`

Reply:
Yes, I *think* you're right. Will think about it more when editing for other things.